include/boost/math/distributions/von_mises.hpp

```diff
@@ -0,0 +1,842 @@
+        BOOST_MATH_BIG_CONSTANT(RealType, 64, +9.82590905134996782086242180e+08),
+        BOOST_MATH_BIG_CONSTANT(RealType, 64, -7.30623197145529889358596301e+09),
+        BOOST_MATH_BIG_CONSTANT(RealType, 64, +3.27310000726207055200805893e+10),
+        BOOST_MATH_BIG_CONSTANT(RealType, 64, -6.64365417189215599168817064e+10)
+    };
+    RealType result = exp(conc * (cos(x - mean) - 1.0));
+    result /= boost::math::tools::evaluate_polynomial(P, RealType(1.0 / conc)) / sqrt(conc)
+              * boost::math::constants::two_pi<RealType>();
+    return result;
+  }
+}
+} // namespace detail
+
+template <class RealType, class Policy>
+inline RealType pdf(const von_mises_distribution<RealType, Policy>& dist, const RealType& x)
+{
+  BOOST_MATH_STD_USING  // for ADL of std functions
+
+  RealType conc = dist.concentration();
+  RealType mean = dist.mean();
+  
+  static const char* function = "boost::math::pdf(const von_mises_distribution<%1%>&, %1%)";
+  
+  RealType result = 0;
+  if (false == detail::check_positive_x(function, conc, &result, Policy()))
+    return result;
+  if (false == detail::check_angle(function, mean, &result, Policy()))
+    return result;
+  if (false == detail::check_angle(function, x - mean, &result, Policy()))
+    return result;
+  // Below produces MSVC 4127 warnings, so the above used instead.
+  //if(std::numeric_limits<RealType>::has_infinity && abs(x) == std::numeric_limits<RealType>::infinity())
+  //{ // pdf + and - infinity is zero.
+  //  return 0;
+  //}
+  typedef boost::integral_constant<int,
+     ((std::numeric_limits<RealType>::digits == 0) || (std::numeric_limits<RealType>::radix != 2)) ?
+     0 :
+     std::numeric_limits<RealType>::digits <= 24 ?
+     24 :
+     std::numeric_limits<RealType>::digits <= 53 ?
+     53 :
+     std::numeric_limits<RealType>::digits <= 64 ?
+     64 :
+     std::numeric_limits<RealType>::digits <= 113 ?
+     113 : -1
+     > tag_type;
+
+  return detail::pdf_impl(dist, x, tag_type());
+} // pdf
+
+namespace detail {
+
+template <class RealType, class Policy>
+inline RealType cdf_impl(const von_mises_distribution<RealType, Policy>& dist, const RealType& x) 
+{
+  BOOST_MATH_STD_USING  // for ADL of std functions
+  
+  RealType conc = dist.concentration();
+  RealType mean = dist.mean();
+  
+  RealType u = x - mean;
+  if (u <= -boost::math::constants::pi<RealType>())
+    return 0;
+  if (u >= +boost::math::constants::pi<RealType>())
+    return 1;  
+
+  // We use the Fortran algorithm designed by Geoffrey W. Hill in
+  // "Algorithm 518: Incomplete Bessel Function I0. The Von Mises Distribution", 1977, ACM
+  // DOI: 10.1145/355744.355753
+  RealType result = 0;
+  
+  int digits = std::numeric_limits<RealType>::max_digits10 - 1;
+  RealType ck = ((0.1611*digits - 2.8778)*digits + 18.45)*digits - 35.4;
+  if (conc > ck) {
+    RealType c = 24.0 * conc;
+    RealType v = c - 56;
+    RealType r = sqrt((54.0 / (347.0 / v + 26.0 - c) - 6.0 + c) / 12.0);    
```

Comment:
Lots of double precision constants here. Is arbitrary precision not possible?

Reply:
As for the CDF, the mentioned paper gives constants depending on the precision for the values of `ck`, `a1`–`a4` and `v`. However, they are only for upto 12 decimal digits. I tried to extrapolate these values but that wasn't quite successful. I have looked around for implementations in other libraries and they all seem to use this old paper and the values for 8 digits (single precision).

Reply:
That's a little disappointing (though probably good enough for most users).

https://www.rdocumentation.org/packages/circular/versions/0.4-93 R documents mention this but I haven't found/downloaded the source code yet, but perhaps you already have?

These fitting functions are a Black Art but I know the Chief Wizard and perhaps he can advise @jzmaddock .

Reply:
I do wonder if we can do better here:

1) I notice you've pulled out and duplicated the contents of I0 - why not just call that function directly?
2) Looking at the original paper, the series to evaluate for the CDF is given on the start of page 280 and everything else is just "faff" to try to evaluate that.  I confess it's going to be a nasty one to code up because the Bessel function recurrences are stable in the backwards not forwards direction.  But it would be true arbitrary precision.

I faced something similar with some of the 1F1 series expansions, and the way I coded it was to maintain a cache of Bessel function values, when filling the cache you start with an arbitrary small value, then iterate to fill the cache and then normalise all the values afterwards either using the last value from the previous cache or a single Bessel function call.  Which is not easy to explain and may or may not make sense!  

The code in question is `hypergeometric_1F1_AS_13_3_7_tricomi_series::refill_cache` in https://github.com/boostorg/math/blob/8eb40afd6b2cdc3ffa786d4494bfc4164794e8d4/include/boost/math/special_functions/detail/hypergeometric_1F1_bessel.hpp#L153.  Note that this is way over-complex for your case because it deals with both Bessel I and J, and positive and negative orders, neither of which are needed here.

Reply:
The reason to duplicate the code was that I0(kappa) grows very fast just like e^kappa, dividing these two however leads to numbers < 100. So, in order to make the code work for large values of kappa, I calculate the quotient directly via polynomial approximation.

Reply:
> The reason to duplicate the code was that I0(kappa) grows very fast just like e^kappa, dividing these two however leads to numbers < 100. So, in order to make the code work for large values of kappa, I calculate the quotient directly via polynomial approximation.

Can you just test for values of kappa that cause overflow and then use an asymptotic there?